File: _CONTAINER/_BY_MONTH/_Feb/Thesis-physics-master/doc/particles/particleSystem2.cpp
/**
 * @fileoverview Particle system.
 * @author sebastian.poreba@gmail.com (Sebastian Poreba)
 */
#include "particleSystem2.h"

smash::ParticleSystem2::ParticleSystem2() {
  this->particles = new std::vector<smash::Particle*>;
  this->deadParticles = new std::stack<smash::Particle*>;
  this->emitters = new std::vector<smash::ParticleEmitter*>;
};

smash::ParticleSystem2::~ParticleSystem2() {
  this->particles->erase(this->particles->begin(), this->particles->end());
  delete this->particles;
  while (!this->deadParticles->empty()) {
    delete this->deadParticles->top();
    this->deadParticles->pop();
  }
  delete this->deadParticles;
  this->emitters->erase(this->emitters->begin(), this->emitters->end());
  delete this->emitters;
};

void smash::ParticleSystem2::step() {
  for (std::vector<smash::ParticleEmitter*>::iterator it =
      this->emitters->begin(); it != this->emitters->end(); it++) {
    smash::ParticleEmitter* emitter = *it;
    for (int i = 0; i < emitter->productionRate; i++) {
      smash::Particle* p;
      if (!this->deadParticles->empty()) {
        p = this->deadParticles->top();
        this->deadParticles->pop();
        p->reset();
      } else {
        p = new smash::Particle();
        this->particles->push_back(p);
      }
      
      p->lifespan = emitter->lifespan;
      p->positionX = emitter->positionX;
      p->positionY = emitter->positionY;
      p->velocityX = sin(emitter->angle +
          (((float) rand() / (RAND_MAX)) - 0.5) * emitter->spread) *
          emitter->velocity * 
          (1 + (((float) rand() / (RAND_MAX)) - 0.5) * emitter->velocitySpread);
      p->velocityY = cos(emitter->angle +
          (((float) rand() / (RAND_MAX)) - 0.5) * emitter->spread) *
          emitter->velocity *
          (1 + (((float) rand() / (RAND_MAX)) - 0.5) * emitter->velocitySpread);
    }
  }

  for (std::vector<smash::Particle*>::iterator it = this->particles->begin();
      it != this->particles->end(); it++) {
    smash::Particle* p = *it;
    if (p->isDead) {
      continue;
    }
    p->step();
    if (p->positionX < 0 ||
        p->positionX >= smash::ParticleSystem2::CANVAS_WIDTH ||
        p->positionY < 0 ||
        p->positionY >= smash::ParticleSystem2::CANVAS_HEIGHT ||
        p->age > p->lifespan) {
      this->deadParticles->push(p);
      p->isDead = true;
    }
  };
};

/**
 * @param emitter
 */
void smash::ParticleSystem2::addEmitter(smash::ParticleEmitter* emitter) {
  this->emitters->push_back(emitter);
};

File: _CONTAINER/_BY_MONTH/_Feb/Thesis-physics-master/browser/engine/particleSystem2.js
/**
 * @fileoverview Particle system.
 * @author sebastian.poreba@gmail.com (Sebastian Poreba)
 */

goog.provide('smash.ParticleSystem2');

goog.require('smash.Particle');
goog.require('smash.flags');



/**
 * @constructor
 */
smash.ParticleSystem2 = function() {
  /**
   * @type {!Array.<smash.Particle>}
   */
  this.particles = [];

  /**
   * @type {!Array.<number>}
   */
  this.deadParticles = [];

  /**
   * @type {!Array.<smash.ParticleEmitter>}
   */
  this.emitters = [];

  if (smash.flags.DRAWING_ENABLED) {
    /**
     * @type {!Element}
     */
    this.canvas = window.document.createElement('canvas');
    this.canvas.width = smash.ParticleSystem2.CANVAS_WIDTH;
    this.canvas.height = smash.ParticleSystem2.CANVAS_HEIGHT;
    window.document.body.appendChild(this.canvas);

    /**
     * @type {!CanvasRenderingContext2D}
     */
    this.context = this.canvas.getContext('2d');

    /**
     * @type {!ImageData}
     */
    this.imageData = this.context.getImageData(0, 0,
        smash.ParticleSystem2.CANVAS_WIDTH,
        smash.ParticleSystem2.CANVAS_HEIGHT);

    /**
     * @type {!CanvasPixelArray}
     */
    this.pixels = this.imageData.data;
  }
};


/**
 * @const {number}
 */
smash.ParticleSystem2.CANVAS_WIDTH = 1200;


/**
 * @const {number}
 */
smash.ParticleSystem2.CANVAS_HEIGHT = 400;


/**
 *
 */
smash.ParticleSystem2.prototype.step = function() {
  if (smash.flags.DRAWING_ENABLED) {
    for (var i = 0; i < smash.ParticleSystem2.CANVAS_WIDTH *
        smash.ParticleSystem2.CANVAS_HEIGHT * 4; i += 4) {
      this.pixels[i] = 0;
      this.pixels[i + 1] = 0;
      this.pixels[i + 2] = 0;
      this.pixels[i + 3] = 0;
    }
  }

  for (var ei = 0; ei < this.emitters.length; ei++) {
    var emitter = this.emitters[ei];
    for (var i = 0; i < emitter.productionRate; i++) {
      var pIndex = this.deadParticles.pop();
      if (pIndex !== undefined) {
        var p = this.particles[pIndex];
        p.reset();
      } else {
        var p = new smash.Particle();
        this.particles.push(p);
      }

      p.lifespan = emitter.lifespan;
      p.positionX = emitter.positionX;
      p.positionY = emitter.positionY;
      p.velocityX = Math.sin(emitter.angle +
          (Math.random() - 0.5) * emitter.spread) *
          emitter.velocity * emitter.velocitySpread;
      p.velocityY = Math.cos(emitter.angle +
          (Math.random() - 0.5) * emitter.spread) *
          emitter.velocity *
          (1 + (Math.random() - 0.5) * emitter.velocitySpread);
    }
  }

  for (var i = 0; i < this.particles.length; i++) {
    var p = this.particles[i];
    p.step();
    if (p.positionX < 0 ||
        p.positionX >= smash.ParticleSystem2.CANVAS_WIDTH ||
        p.positionY < 0 ||
        p.positionY >= smash.ParticleSystem2.CANVAS_HEIGHT ||
        p.age > p.lifespan) {
      this.deadParticles.push(i);
      p.isDead = true;
    }

    if (smash.flags.DRAWING_ENABLED && !p.isDead) {
      var baseIndex =
          (Math.round(p.positionY) *
              smash.ParticleSystem2.CANVAS_WIDTH +
              Math.round(p.positionX)) * 4;
      this.pixels[baseIndex] = Math.round(p.velocityX * 80);
      this.pixels[baseIndex + 1] = Math.round(p.velocityX * 80);
      this.pixels[baseIndex + 2] = 255 - Math.round(p.age / p.lifespan * 255);
      this.pixels[baseIndex + 3] = 255;
    }
  }

  if (smash.flags.DRAWING_ENABLED) {
    this.context.putImageData(this.imageData, 0, 0);
  }
};


/**
 * @param {!smash.ParticleEmitter} emitter
 */
smash.ParticleSystem2.prototype.addEmitter = function(emitter) {
  this.emitters.push(emitter);
};
